Improve buildchain verbosity and readability

alexandre-allard opened the pull request:
**Component**: buildchain

**Context**:
In the buildchain, we are actually missing some output for debugging purposes and some output are not really readable.
We also do not have timers to see the time taken by each step.

**Summary**:
* Clean up the docker build//run output on failure (make it more readable)
* Add timer for each step

I still need to basically pipe every logs to a file.

**Acceptance criteria**: 

---

Closes: #3217

alexandre-allard committed:
build: Clean up docker run output

Previously we were printing the error from
a ran command 4 times and with all new lines
escaped which makes the output not really
readable.
We now print the error only once and eval
the return from docker run to interpret
the new lines.

Refs: #3217

alexandre-allard committed:
build: Clean up docker build error output

Only display once the failed command

Refs: #3217

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

alexandre-allard committed:
build: Display the time elapsed for tasks

Each task not skipped or ignored is now print
2 times when started and ended, whether it succeed
or not.
This is necessary to display the time elapsed
even with multithreaded builds.
Also change the status of each step to more
human readable status and colors.

Refs: #3217

gdemonet commented on buildchain/dodo.py:
```diff
@@ -112,6 +120,11 @@ def _write_failure(self, result: Dict[str, Any]) -> None:
 
+    def _write_time_elapsed(self, task: doit.task.Task) -> None:
+        if task.actions:
+            time_elapsed = int(self.t_results[task.name].to_dict()["elapsed"])
```
`TaskResult.elapsed` should be of type `Optional[float]` already, so you're using the type cast as an implicit guard against `None` values? Or just to essentially truncate the float value to remove sub-second precision digits? Either way, could be better to be explicit, either with a comment or with more explicit code:

```python
# Use `to_dict` to evaluate the "elapsed" value
task_result = self.t_results[task.name].to_dict()
elapsed = task_result["elapsed"]
assert elapsed is not None, "Some runtime error message"
# `elapsed` is a float, but we strip sub-second digits
self._write(f" [{elapsed:.0f}s]\n")
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

alexandre-allard committed:
build: Add CmdActionOnFailure class

It adds a parameter `on_failure` to `CmdAction`
which allows to run any callable when the command
fails (useful to clean up on failure).
Replace subprocess.run by this class, this way
we benefits from the stdout/stderr capture
system, etc.

Refs: #3217

alexandre-allard committed:
build: use patched version of doit

We target a specific commit hash for
doit package as we need a fix not released yet.
The doit version should be changed once a version
including it is released.

alexandre-allard committed:
build: Fix docker build error

Error happening during the build phase
of a container where always throwing a
traceback since we were trying to access
to non-existing keys.
Also improve a little bit this output
to be more like the Docker CLI one.

alexandre-allard committed:
build: Make correct use of verbosity and failure-verbosity options

We can now adapt the output verbosity using
--verbosity and --failure-verbosity options
as doit documents it.

Refs: #3217

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/3217-buildchain-verbosity` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `improvement/3217-buildchain-verbosity`
* Rebase `improvement/3217-buildchain-verbosity` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

alexandre-allard committed:
eve: raise doit verbosity in CI

Refs: #3217

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/3217-buildchain-verbosity` and the
destination branch `development/2.9`.

Please resolve the conflict on **the feature branch** (`improvement/3217-buildchain-verbosity`).

```bash
 $ git fetch
 $ git checkout origin/improvement/3217-buildchain-verbosity
 $ git merge origin/development/2.9
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/3217-buildchain-verbosity
```

TeddyAndrieux commented on buildchain/buildchain/action.py:
```diff
@@ -0,0 +1,36 @@
+import doit  # type: ignore
+
+
+class CmdActionOnFailure(doit.action.CmdAction):  # type: ignore
+    """Wrapper around CmdAction allowing to pass a callable run if action fails"""
+
+    def __init__(
+        self,
+        action: Union[str, List[Union[str, Path]], Callable[..., Optional[Any]]],
+        task: Optional[doit.task.Task] = None,
+        save_out: Optional[str] = None,
+        shell: bool = True,
+        encoding: str = "utf-8",
+        decode_error: str = "replace",
+        buffering: int = 0,
+        on_failure: Optional[Callable[..., None]] = None,
+        **pkwargs: Any
+    ):
+        super().__init__(
+            action, task, save_out, shell, encoding, decode_error, buffering, **pkwargs
+        )
+        self.on_failure = on_failure
+
+    def execute(
+        self, out: Any = None, err: Any = None
+    ) -> doit.exceptions.CatchedException:
+        failure = super().execute(out, err)
+        if failure and self.on_failure:
+            self.on_failure()
+        return failure
```
Maybe it's also something that can be part of upstream doit ? (just a question not said it should be done right now)

alexandre-allard replied:
Yes, maybe, but since anyone can easily overload this class, it's not an issue (the other fix was necessary since it touches a part that can't be overloaded).

TeddyAndrieux replied:
Yes sure it's not needed

TeddyAndrieux commented on tox.ini:
```diff
@@ -23,7 +23,7 @@ setenv =
     bash -c ' \
         pip-compile \
-            --emit-index-url --emit-trusted-host --annotate --generate-hashes \
+            --emit-index-url --emit-trusted-host --annotate \
```
Why ? It does not work when using `git` with a hash ?

alexandre-allard replied:
Yes

TeddyAndrieux commented on buildchain/buildchain/docker_command.py:
```diff
@@ -63,11 +63,18 @@ def build_error_handler(build_error: BuildError) -> str:
+                    item["progressDetail"]["current"],
+                    item["progressDetail"]["total"],
+                )
+            except KeyError:
+                line += "\n"
```
Not sure to understand why we only add a break line if we do not have progressDetail

alexandre-allard replied:
If we display progressDetail, we want it to be behave like it's done on CLI output, re-writring previous progress line with the new one, otherwise we want to add a newline for the sake of readability (to avoid having everything on a single line).
To help you better understand, here is an example of what's in the logs, each line being an `item`:
```
{'status': 'Pulling from library/centos', 'id': 'sha256:5d4f4e6051c7cc10f2e712f9dc3f86a2bd67e457bced7ca52a71c243099c0121'}
{'status': 'Pulling fs layer', 'progressDetail': {}, 'id': '581e78aaf612'}
{'status': 'Downloading', 'progressDetail': {'current': 73603419, 'total': 75167150}, 'progress': '[================================================>  ]   73.6MB/75.17MB', 'id': '581e78aaf612'}
[...]
{'status': 'Downloading', 'progressDetail': {'current': 74680667, 'total': 75167150}, 'progress': '[=================================================> ]  74.68MB/75.17MB', 'id': '581e78aaf612'}
{'status': 'Verifying Checksum', 'progressDetail': {}, 'id': '581e78aaf612'}
{'status': 'Download complete', 'progressDetail': {}, 'id': '581e78aaf612'}
{'status': 'Extracting', 'progressDetail': {'current': 557056, 'total': 75167150}, 'progress': '[>                                                  ]  557.1kB/75.17MB', 'id': '581e78aaf612'}
[...]
{'status': 'Extracting', 'progressDetail': {'current': 69074944, 'total': 75167150}, 'progress': '[=============================================>     ]  69.07MB/75.17MB', 'id': '581e78aaf612'}
{'status': 'Pull complete', 'progressDetail': {}, 'id': '581e78aaf612'}
{'status': 'Digest: sha256:5d4f4e6051c7cc10f2e712f9dc3f86a2bd67e457bced7ca52a71c243099c0121'}
{'status': 'Status: Downloaded newer image for centos@sha256:5d4f4e6051c7cc10f2e712f9dc3f86a2bd67e457bced7ca52a71c243099c0121'}
```

alexandre-allard replied:
We could also not display this progressDetail at all, since normally it's overwritten by the next line (e.g. 'Verifying Checksum'), it's just in case the download/extract would be stopped in the middle.

TeddyAndrieux replied:
Oh ok, got it, thank you for the explanation :ok_hand:

TeddyAndrieux reviewed (dismissed):
Few minor comments/questions, otherwise LGTM

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

alexandre-allard commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
